zig implementation ported to odin

temidaradev opened the pull request:
Test results:
```bash
[INFO ] --- [2025-06-13 09:38:56] Starting test runner with 1 thread. Set with -define:ODIN_TEST_THREADS=n.
[INFO ] --- [2025-06-13 09:38:56] The random seed sent to every test is: 8438900958400. Set with -define:ODIN_TEST_RANDOM_SEED=n.
[INFO ] --- [2025-06-13 09:38:56] Memory tracking is enabled. Tests will log their memory usage if there's an issue.
[INFO ] --- [2025-06-13 09:38:56] < Final Mem/ Total Mem> <  Peak Mem> (#Free/Alloc) :: [package.test_name]
[WARN ] --- [2025-06-13 09:38:56] <   2.73KiB/   3.21KiB> <   3.03KiB> (    3/   22) :: quakemap.test_quakemap_read
        +++ leak        48B @ 0xC928006F8 [geometry.odin:64:compute_vertices()]
        +++ leak       256B @ 0xC92800958 [parser.odin:105:read_entity()]
        +++ leak       768B @ 0xC928002B0 [parser.odin:214:read_solid()]
        +++ leak         6B @ 0xC92800939 [parser.odin:102:read_entity()]
        +++ leak        48B @ 0xC928007A0 [geometry.odin:64:compute_vertices()]
        +++ leak         8B @ 0xC92800947 [parser.odin:103:read_entity()]
        +++ leak        15B @ 0xC9280018D [parser.odin:103:read_entity()]
        +++ leak        48B @ 0xC928006C0 [geometry.odin:64:compute_vertices()]
        +++ leak       256B @ 0xC928001A8 [parser.odin:105:read_entity()]
        +++ leak        11B @ 0xC92800A60 [parser.odin:102:read_entity()]
        +++ leak       832B @ 0xC92800A80 [parser.odin:42:read()]
        +++ leak        48B @ 0xC92800768 [geometry.odin:64:compute_vertices()]
        +++ leak        48B @ 0xC92800688 [geometry.odin:64:compute_vertices()]
        +++ leak        10B @ 0xC92800170 [parser.odin:92:read_entity()]
        +++ leak        48B @ 0xC92800730 [geometry.odin:64:compute_vertices()]
        +++ leak         3B @ 0xC92800182 [parser.odin:102:read_entity()]
        +++ leak        17B @ 0xC92800920 [parser.odin:92:read_entity()]
        +++ leak       320B @ 0xC928007D8 [parser.odin:114:read_entity()]
        +++ leak         5B @ 0xC92800A73 [parser.odin:103:read_entity()]
quakemap  [|                       ]         1 :: [package done]

Finished 1 test in 217µs. The test was successful.
```

temidaradev committed:
zig implementation ported to odin

temidaradev committed:
mem leaks fixed

temidaradev commented:
```bash
[INFO ] --- [2025-06-13 10:02:04] Starting test runner with 1 thread.
[INFO ] --- [2025-06-13 10:02:04] The random seed sent to every test is: 9826725181150. Set with -define:ODIN_TEST_RANDOM_SEED=n.
[INFO ] --- [2025-06-13 10:02:04] Memory tracking is enabled. Tests will log their memory usage if there's an issue.
[INFO ] --- [2025-06-13 10:02:04] < Final Mem/ Total Mem> <  Peak Mem> (#Free/Alloc) :: [package.test_name]
quakemap  [|                       ]         1 :: [package done]

Finished 1 test in 276µs. The test was successful.
```

temidaradev committed:
basic test environment without textures

temidaradev committed:
textures :o

temidaradev commented on src/quakemap/geometry.odin:
```diff
@@ -1,26 +1,163 @@
+			t := sd / (sd - ed)
+			intersect := linalg.lerp(s, e, t)
+
+			// Snap to plane if normal is axis-aligned
+			if clip_plane.normal.x == 1 do intersect.x = -clip_plane.d
+			if clip_plane.normal.x == -1 do intersect.x = clip_plane.d
+			if clip_plane.normal.y == 1 do intersect.y = -clip_plane.d
+			if clip_plane.normal.y == -1 do intersect.y = clip_plane.d
+			if clip_plane.normal.z == 1 do intersect.z = -clip_plane.d
+			if clip_plane.normal.z == -1 do intersect.z = clip_plane.d
```
bruh gay

temidaradev commented on src/quakemap/geometry.odin:
```diff
@@ -1,26 +1,163 @@
+				return .Degenerate_Face
+			}
+
+			// Swap vertices and clipped
+			vertices, clipped = clipped, vertices
```
This reintroduces copying and potentially reallocations, increasing overhead. gay

temidaradev commented:
Currently gets random texture from the textures folder and all stuff gets covered by selected texture.

https://github.com/user-attachments/assets/8fe0b2d0-c06e-4c86-9aaa-ed252930a967

nii236 commented on TrenchBroomTest/raylib_viewer:
Add this to .gitignore, we don't want compiled binaries in the repo

nii236 commented on TrenchBroomTest/raylib_viewer.odin:
```diff
@@ -0,0 +1,471 @@
+	fmt.printf("Converted %d meshes to Raylib\n", len(raylib_meshes))
+}
+
+convert_mesh_to_raylib :: proc(quake_mesh: quakemap.Mesh) -> rl.Mesh {
```
Whats difference between this proc and  `convert_quake_meshes_to_raylib`? Names are similar

maybe this should be in the quakemap lib too?

nii236 replied:
This should return an error enum

nii236 commented on TrenchBroomTest/raylib_viewer.odin:
```diff
@@ -0,0 +1,471 @@
+	}
+
+	// Upload mesh data to GPU
+	rl.UploadMesh(&mesh, false)
```
this should be in the calling proc

nii236 commented on TrenchBroomTest/raylib_viewer.odin:
```diff
@@ -0,0 +1,471 @@
+	}
+}
+
+draw_collision_wireframes :: proc() {
```
this isn't used anywhere?

nii236 commented on TrenchBroomTest/raylib_viewer.odin:
```diff
@@ -0,0 +1,471 @@
+	}
+}
+
+draw_spawn_points :: proc() {
```
same

whats with the dead code slop? you need to check the ai output more carefully before i look at it

temidaradev replied:
i was testing so i forgot it in there .d

nii236 commented on TrenchBroomTest/raylib_viewer.odin:
```diff
@@ -0,0 +1,471 @@
+	}
+}
+
+draw_ui :: proc() {
```
same

temidaradev replied:
no this is used for crosshair

nii236 commented on TrenchBroomTest/raylib_viewer.odin:
```diff
@@ -0,0 +1,471 @@
+
+// Extract material name from a quakemap mesh
+// This is a simplified approach - we'll use the first texture that matches the dimensions
+// In a real implementation, we'd want the quakemap library to store the material name
```
implement textures properly in the library as commented here

nii236 replied:
only 1 texture is rendered. where are the others?

<img width="1312" alt="image" src="https://github.com/user-attachments/assets/27c0dc68-7f8f-4183-9412-4750e0c3d62c" />

nii236 commented on TrenchBroomTest/raylib_viewer.odin:
```diff
@@ -0,0 +1,471 @@
+	camera_pitch = 0.0
+}
+
+load_textures :: proc() {
```
This should return an error enum

nii236 commented on TrenchBroomTest/raylib_viewer.odin:
```diff
@@ -0,0 +1,471 @@
+	fmt.printf("Loaded %d textures\n", len(loaded_textures))
+}
+
+convert_quake_meshes_to_raylib :: proc() {
```
This should return an error enum

nii236 commented on TrenchBroomTest/raylib_viewer.odin:
```diff
@@ -0,0 +1,471 @@
+	rl.DrawLine(center_x, center_y - 10, center_x, center_y + 10, rl.WHITE)
+}
+
+// Extract material name from a quakemap mesh
```
This should return an error enum

nii236 commented on TrenchBroomTest/raylib_viewer.odin:
```diff
@@ -0,0 +1,471 @@
+setup_camera :: proc() {
+	initial_pos := rl.Vector3{0, 10, 0}
+
+	if len(loaded_map.spawn_points) > 0 {
```
return error if loaded_map is invalid

nii236 reviewed (changes requested):
lots of slop in here, unused functions or not returning errors

nii236 commented:
Rename trenchbroomtest package and folder to `example`

temidaradev commented:
oops looks like i forgot to delete old codes

temidaradev committed:
rename folder to example and ignore binary

temidaradev committed:
use more textures

temidaradev committed:
consistent texture usage and using material name

temidaradev committed:
convert_mesh_to_raylib returns error enum

temidaradev committed:
The call to rl.UploadMesh(&mesh, false) is now moved out of convert_mesh_to_raylib

temidaradev committed:
unused codes removed

temidaradev committed:
The load_textures procedure now returns a LoadTexturesError enum error

temidaradev committed:
The convert_quake_meshes_to_raylib procedure now returns a ConvertMeshesError enum error

temidaradev committed:
convert_quake_meshes_to_raylib is now build_raylib_models_from_map
convert_mesh_to_raylib is now build_raylib_mesh_from_quakemap and important error enums added

temidaradev committed:
os.exit(1) if loaded_map is invalid

temidaradev commented:
Here how they look
https://github.com/user-attachments/assets/20a151ae-3de7-4871-bbae-aad80f4270f5

temidaradev commented:
```bash
=== Raylib Quake Map Viewer ===
INFO: Initializing raylib 5.5
INFO: Platform backend: DESKTOP (GLFW)
INFO: Supported raylib modules:
INFO:     > rcore:..... loaded (mandatory)
INFO:     > rlgl:...... loaded (mandatory)
INFO:     > rshapes:... loaded (optional)
INFO:     > rtextures:. loaded (optional)
INFO:     > rtext:..... loaded (optional)
INFO:     > rmodels:... loaded (optional)
INFO:     > raudio:.... loaded (optional)
INFO: DISPLAY: Device initialized successfully
INFO:     > Display size: 1920 x 1080
INFO:     > Screen size:  1200 x 800
INFO:     > Render size:  1200 x 800
INFO:     > Viewport offsets: 0, 0
INFO: GLAD: OpenGL extensions loaded successfully
INFO: GL: Supported extensions count: 242
INFO: GL: OpenGL device information:
INFO:     > Vendor:   AMD
INFO:     > Renderer: AMD Radeon RX 6600 (radeonsi, navi23, LLVM 20.1.6, DRM 3.63, 6.15.2-arch1-1)
INFO:     > Version:  4.6 (Core Profile) Mesa 25.1.3-arch1.3
INFO:     > GLSL:     4.60
INFO: GL: VAO extension detected, VAO functions loaded successfully
INFO: GL: NPOT textures extension detected, full NPOT textures supported
INFO: GL: DXT compressed textures supported
INFO: GL: ETC2/EAC compressed textures supported
INFO: PLATFORM: DESKTOP (GLFW - X11): Initialized successfully
INFO: TEXTURE: [ID 1] Texture loaded successfully (1x1 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 1] Default texture loaded successfully
INFO: SHADER: [ID 1] Vertex shader compiled successfully
INFO: SHADER: [ID 2] Fragment shader compiled successfully
INFO: SHADER: [ID 3] Program shader loaded successfully
INFO: SHADER: [ID 3] Default shader loaded successfully
INFO: RLGL: Render batch vertex buffers loaded successfully in RAM (CPU)
INFO: RLGL: Render batch vertex buffers loaded successfully in VRAM (GPU)
INFO: RLGL: Default OpenGL state initialized successfully
INFO: TEXTURE: [ID 2] Texture loaded successfully (128x128 | GRAY_ALPHA | 1 mipmaps)
INFO: FONT: Default font loaded successfully (224 glyphs)
INFO: SYSTEM: Working Directory: /home/temidaradev/Documents/quakemap-odin/example
DEBUG: Starting load_textures
DEBUG: Opening texture directory: textures
DEBUG: Directory opened successfully
DEBUG: Found 15 files in directory
DEBUG: Processing file: tech_1.png
DEBUG: Loading texture: tech_1 from textures/tech_1.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_1.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 3] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 3
Loaded texture: tech_1
DEBUG: Processing file: tech_11.png
DEBUG: Loading texture: tech_11 from textures/tech_11.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_11.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 4] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 4
Loaded texture: tech_11
DEBUG: Processing file: tech_13.png
DEBUG: Loading texture: tech_13 from textures/tech_13.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_13.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 5] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 5
Loaded texture: tech_13
DEBUG: Processing file: tech_14.png
DEBUG: Loading texture: tech_14 from textures/tech_14.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_14.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 6] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 6
Loaded texture: tech_14
DEBUG: Processing file: tech_15.png
DEBUG: Loading texture: tech_15 from textures/tech_15.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_15.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 7] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 7
Loaded texture: tech_15
DEBUG: Processing file: tech_16.png
DEBUG: Loading texture: tech_16 from textures/tech_16.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_16.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 8] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 8
Loaded texture: tech_16
DEBUG: Processing file: tech_18.png
DEBUG: Loading texture: tech_18 from textures/tech_18.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_18.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 9] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 9
Loaded texture: tech_18
DEBUG: Processing file: tech_2.png
DEBUG: Loading texture: tech_2 from textures/tech_2.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_2.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 10] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 10
Loaded texture: tech_2
DEBUG: Processing file: tech_3.png
DEBUG: Loading texture: tech_3 from textures/tech_3.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_3.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 11] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 11
Loaded texture: tech_3
DEBUG: Processing file: tech_4.png
DEBUG: Loading texture: tech_4 from textures/tech_4.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_4.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 12] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 12
Loaded texture: tech_4
DEBUG: Processing file: tech_5.png
DEBUG: Loading texture: tech_5 from textures/tech_5.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_5.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 13] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 13
Loaded texture: tech_5
DEBUG: Processing file: tech_6.png
DEBUG: Loading texture: tech_6 from textures/tech_6.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_6.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 14] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 14
Loaded texture: tech_6
DEBUG: Processing file: tech_7.png
DEBUG: Loading texture: tech_7 from textures/tech_7.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_7.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 15] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 15
Loaded texture: tech_7
DEBUG: Processing file: tech_8.png
DEBUG: Loading texture: tech_8 from textures/tech_8.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_8.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 16] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 16
Loaded texture: tech_8
DEBUG: Processing file: tech_9.png
DEBUG: Loading texture: tech_9 from textures/tech_9.png
DEBUG: About to call rl.LoadTexture
INFO: FILEIO: [textures/tech_9.png] File loaded successfully
INFO: IMAGE: Data loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 17] Texture loaded successfully (32x32 | R8G8B8A8 | 1 mipmaps)
DEBUG: rl.LoadTexture returned, texture.id = 17
Loaded texture: tech_9
Loaded 15 textures in sorted order
Registered material: tech_8 (32x32)
Registered material: tech_4 (32x32)
Registered material: tech_16 (32x32)
Registered material: tech_13 (32x32)
Registered material: tech_18 (32x32)
Registered material: tech_6 (32x32)
Registered material: tech_1 (32x32)
Registered material: tech_14 (32x32)
Registered material: tech_11 (32x32)
Registered material: tech_3 (32x32)
Registered material: tech_9 (32x32)
Registered material: tech_5 (32x32)
Registered material: tech_15 (32x32)
Registered material: tech_7 (32x32)
Registered material: tech_2 (32x32)
Loading map file: test_empty.map
Map loaded successfully!
INFO: TIMER: Target time per frame: 16.667 milliseconds
Starting at spawn point: [0, 0, 24] (scaled: [0, 3.4, -0])
INFO: VAO: [ID 2] Mesh uploaded successfully to VRAM (GPU)
DEBUG: No material handle in mesh
INFO: VAO: [ID 3] Mesh uploaded successfully to VRAM (GPU)
DEBUG: Material name found via handle: 'tech_6'
INFO: VAO: [ID 4] Mesh uploaded successfully to VRAM (GPU)
DEBUG: Material name found via handle: 'tech_13'
INFO: VAO: [ID 5] Mesh uploaded successfully to VRAM (GPU)
DEBUG: Material name found via handle: 'tech_14'
INFO: VAO: [ID 6] Mesh uploaded successfully to VRAM (GPU)
DEBUG: Material name found via handle: 'tech_7'
INFO: VAO: [ID 7] Mesh uploaded successfully to VRAM (GPU)
DEBUG: No material handle in mesh
INFO: VAO: [ID 8] Mesh uploaded successfully to VRAM (GPU)
DEBUG: No material handle in mesh
INFO: VAO: [ID 9] Mesh uploaded successfully to VRAM (GPU)
DEBUG: Material name found via handle: 'tech_2'
INFO: VAO: [ID 10] Mesh uploaded successfully to VRAM (GPU)
DEBUG: Material name found via handle: 'tech_11'
INFO: VAO: [ID 11] Mesh uploaded successfully to VRAM (GPU)
DEBUG: Material name found via handle: 'tech_9'
INFO: VAO: [ID 12] Mesh uploaded successfully to VRAM (GPU)
DEBUG: Material name found via handle: 'tech_3'
Starting render loop...
INFO: VAO: [ID 2] Unloaded vertex array data from VRAM (GPU)
INFO: MODEL: Unloaded model (and meshes) from RAM and VRAM
INFO: VAO: [ID 3] Unloaded vertex array data from VRAM (GPU)
INFO: MODEL: Unloaded model (and meshes) from RAM and VRAM
INFO: VAO: [ID 4] Unloaded vertex array data from VRAM (GPU)
INFO: MODEL: Unloaded model (and meshes) from RAM and VRAM
INFO: VAO: [ID 5] Unloaded vertex array data from VRAM (GPU)
INFO: MODEL: Unloaded model (and meshes) from RAM and VRAM
INFO: VAO: [ID 6] Unloaded vertex array data from VRAM (GPU)
INFO: MODEL: Unloaded model (and meshes) from RAM and VRAM
INFO: VAO: [ID 7] Unloaded vertex array data from VRAM (GPU)
INFO: MODEL: Unloaded model (and meshes) from RAM and VRAM
INFO: VAO: [ID 8] Unloaded vertex array data from VRAM (GPU)
INFO: MODEL: Unloaded model (and meshes) from RAM and VRAM
INFO: VAO: [ID 9] Unloaded vertex array data from VRAM (GPU)
INFO: MODEL: Unloaded model (and meshes) from RAM and VRAM
INFO: VAO: [ID 10] Unloaded vertex array data from VRAM (GPU)
INFO: MODEL: Unloaded model (and meshes) from RAM and VRAM
INFO: VAO: [ID 11] Unloaded vertex array data from VRAM (GPU)
INFO: MODEL: Unloaded model (and meshes) from RAM and VRAM
INFO: VAO: [ID 12] Unloaded vertex array data from VRAM (GPU)
INFO: MODEL: Unloaded model (and meshes) from RAM and VRAM
INFO: VAO: [ID 2] Unloaded vertex array data from VRAM (GPU)
```

temidaradev commented on src/quakemap/geometry.odin:
```diff
@@ -1,26 +1,163 @@
+			intersect := linalg.lerp(s, e, t)
+
+			// Snap to plane if normal is axis-aligned
+			if clip_plane.normal.x == 1 do intersect.x = -clip_plane.d
+			if clip_plane.normal.x == -1 do intersect.x = clip_plane.d
+			if clip_plane.normal.y == 1 do intersect.y = -clip_plane.d
+			if clip_plane.normal.y == -1 do intersect.y = clip_plane.d
+			if clip_plane.normal.z == 1 do intersect.z = -clip_plane.d
+			if clip_plane.normal.z == -1 do intersect.z = clip_plane.d
```
no
